core/src/main/scala/higherkindness/mu/rpc/srcgen/avro/AvroSrcGenerator.scala

```diff
@@ -138,8 +135,9 @@ final case class AvroSrcGenerator(
         try comment ++ Seq(parseMessage(name, message.getRequest, message.getResponse), "")
         catch {
           case ParseException(msg) =>
-            logger.warn(s"$msg, cannot be converted to mu: $message")
-            Seq.empty
+            throw new RuntimeException(
+              s"$msg; the following protocol cannot be converted to mu: $message"
```

Comment:
I don't love this approach because while it does successfully prevent the plugin from generating any code if it encounters an Avro entry that is not a record, it doesn't handle the failure gracefully.  It coughs up an ugly RuntimeException and it fails the command immediately, and while it does report what the problem was clearly, it behaves like a parser, rather than a compiler.  

The ultimate goal of these changes is to make it so that this plugin reports all of the parsing failures it encounters and then fails at the very end, so that the user can see where _all_ of their issues are, rather than just having to run this step and have it fail each time until they've cleaned up all of their primitive types.  However, to do this, we'd need to change some the contract in our `generateFrom` method so that we can return the errors collected during parsing in addition to outputting the parsed code.

In looking at examples, my favorite one that I've found is `sbt-scalafmt`, which uses a [scalaFmtAnalysis](https://github.com/scalameta/sbt-scalafmt/blob/master/plugin/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala#L69) case class to model the returned results from running a formatter on a file, and then passes that case class into the plugin to make decisions with it (it has an awesome failure handler called [trueOrBoom](https://github.com/scalameta/sbt-scalafmt/blob/a3d9b66e4f6265745447e8e0521eaa1973fedf5f/plugin/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala#L223).  It's an elegant way to track all of the potential errors returned from analyzing a series of files and then reporting them at the very end.  Considering our tool is a parsing and a source-code generator, I like the idea of our tool doing something similar.  Another good example of reporting on the specific error collection is in [sbt-scalafix](https://github.com/scalacenter/sbt-scalafix/blob/master/src/main/scala/scalafix/sbt/ScalafixPlugin.scala#L555).

Now, both of these approaches are more elegant and more useful to the end user than what I've done here, but the downside is that they're slightly more involved and they'd require a more significant change to how we report on results for our existing plugin.  I'm game to take a crack at it, but I wanted to get this quick-n-dirty implementation out first to start that conversation.  Furthermore, I'm wondering if anyone else who's worked on sbt plugins before has more insights on how to collect and report or failures, since my only inspiration was looking around the internet.

Reply:
I agree that accumulating errors makes sense here. I would suggest:

* get rid of ParseException
* instead of throwing exceptions, make `parseMessage` return `ValidatedNel[String, String]`. So it either returns a successful method signature or one or more error messages.
* while you're there, rename `parseMessage`? It's not parsing, it's actually doing the opposite! i.e. turning structured data into a string. `buildMethodSignature`?
* rewrite the `requestLines` bit to use a `ValidatedNel` as well, accumulating errors as it iterates over the messages (I guess this will involve `traverse` or maybe `flatTraverse`)
* if there are any errors, throw a single exception at the end and print them all out